### ISSUE TITLE
Occasional test failures with split Include

### ISSUE THREAD

**Issue by ajcvickers** [dotnet/efcore#21296]
@JeremyLikness encountered this:

```
System.InvalidOperationException : Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct.

   at System.Collections.Generic.Dictionary`2.FindValue(TKey key)
   at System.Collections.Generic.Dictionary`2.TryGetValue(TKey key, TValue& value)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.IdentityMap`1.TryGetEntry(Object[] keyValues, Boolean throwOnNullKey, Boolean& hasNullKey) in C:\Users\jeliknes\source\repos\efcore\src\EFCore\ChangeTracking\Internal\IdentityMap.cs:line 142
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.TryGetEntry(IKey key, Object[] keyValues, Boolean throwOnNullKey, Boolean& hasNullKey) in C:\Users\jeliknes\source\repos\efcore\src\EFCore\ChangeTracking\Internal\StateManager.cs:line 389
   at Microsoft.EntityFrameworkCore.Query.QueryContext.TryGetEntry(IKey key, Object[] keyValues, Boolean throwOnNullKey, Boolean& hasNullKey) in C:\Users\jeliknes\source\repos\efcore\src\EFCore\Query\QueryContext.cs:line 156
   at lambda_method361847(Closure , QueryContext , DbDataReader , ResultContext , SplitQueryResultCoordinator )
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.PopulateSplitIncludeCollectionAsync[TIncludingEntity,TIncludedEntity](Int32 collectionId, RelationalQueryContext queryContext, IExecutionStrategy executionStrategy, SplitQueryResultCoordinator resultCoordinator, RelationalCommandCache relationalCommandCache, Func`3 childIdentifier, IReadOnlyList`1 identifierValueComparers, Func`5 innerShaper, Func`4 relatedDataLoaders, INavigation inverseNavigation, Action`2 fixup, Boolean trackingQuery) in C:\Users\jeliknes\source\repos\efcore\src\EFCore.Relational\Query\RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs:line 1409
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.TaskAwaiter(Task[] tasks) in C:\Users\jeliknes\source\repos\efcore\src\EFCore.Relational\Query\RelationalShapedQueryCompilingExpressionVisitor.ShaperProcessingExpressionVisitor.cs:line 1734
   at Microsoft.EntityFrameworkCore.Query.Internal.SplitQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync() in C:\Users\jeliknes\source\repos\efcore\src\EFCore.Relational\Query\Internal\SplitQueryingEnumerable.cs:line 264
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken) in C:\Users\jeliknes\source\repos\efcore\src\EFCore\Extensions\EntityFrameworkQueryableExtensions.cs:line 2329
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken) in C:\Users\jeliknes\source\repos\efcore\src\EFCore\Extensions\EntityFrameworkQueryableExtensions.cs:line 2329
   at Microsoft.EntityFrameworkCore.TestUtilities.QueryAsserter.AssertQuery[TResult](Func`2 actualQuery, Func`2 expectedQuery, Func`2 elementSorter, Action`2 elementAsserter, Boolean assertOrder, Int32 entryCount, Boolean async, String testMethodName) in C:\Users\jeliknes\source\repos\efcore\test\EFCore.Specification.Tests\TestUtilities\QueryAsserter.cs:line 98
   at Microsoft.EntityFrameworkCore.Query.NorthwindSplitIncludeQuerySqlServerTest.Include_duplicate_collection_result_operator(Boolean async) in C:\Users\jeliknes\source\repos\efcore\test\EFCore.SqlServer.FunctionalTests\Query\NorthwindSplitIncludeQuerySqlServerTest.cs:line 849
--- End of stack trace from previous location ---
```

I have seen this:
```
Microsoft.EntityFrameworkCore.Query.NorthwindSplitIncludeQuerySqlServerTest.Include_duplicate_collection_result_operator(async: True)
Assert.Equal() Failure\nExpected: 15\nActual:   16
   at Microsoft.EntityFrameworkCore.TestUtilities.QueryAsserter.AssertEntryCount(DbContext context, Int32 entryCount) in /home/ajcvickers/efcore/test/EFCore.Specification.Tests/TestUtilities/QueryAsserter.cs:line 1605
   at Microsoft.EntityFrameworkCore.TestUtilities.QueryAsserter.AssertQuery[TResult](Func`2 actualQuery, Func`2 expectedQuery, Func`2 elementSorter, Action`2 elementAsserter, Boolean assertOrder, Int32 entryCount, Boolean async, String testMethodName) in /home/ajcvickers/efcore/test/EFCore.Specification.Tests/TestUtilities/QueryAsserter.cs:line 126
   at Microsoft.EntityFrameworkCore.Query.NorthwindSplitIncludeQuerySqlServerTest.Include_duplicate_collection_result_operator(Boolean async) in /home/ajcvickers/efcore/test/EFCore.SqlServer.FunctionalTests/Query/NorthwindSplitIncludeQuerySqlServerTest.cs:line 849
--- End of stack trace from previous location ---
```

**Comment by ajcvickers** [comment]
Another one:
```
Microsoft.EntityFrameworkCore.Query.NorthwindSplitIncludeQuerySqlServerTest.Include_duplicate_collection(async: True)
Assert.Equal() Failure\nExpected: 34\nActual:   35
   at Microsoft.EntityFrameworkCore.TestUtilities.QueryAsserter.AssertEntryCount(DbContext context, Int32 entryCount) in /home/ajcvickers/efcore/test/EFCore.Specification.Tests/TestUtilities/QueryAsserter.cs:line 1605
   at Microsoft.EntityFrameworkCore.TestUtilities.QueryAsserter.AssertQuery[TResult](Func`2 actualQuery, Func`2 expectedQuery, Func`2 elementSorter, Action`2 elementAsserter, Boolean assertOrder, Int32 entryCount, Boolean async, String testMethodName) in /home/ajcvickers/efcore/test/EFCore.Specification.Tests/TestUtilities/QueryAsserter.cs:line 126
   at Microsoft.EntityFrameworkCore.Query.NorthwindSplitIncludeQuerySqlServerTest.Include_duplicate_collection(Boolean async) in /home/ajcvickers/efcore/test/EFCore.SqlServer.FunctionalTests/Query/NorthwindSplitIncludeQuerySqlServerTest.cs:line 789
--- End of stack trace from previous location ---
```

**Comment by smitpatel** [comment]
@AndriySvyryd figured out the issue in the generated shaper which was basically causing a race condition.
I learned something new today. Passing `new Task [] { Populate(), Populate()}` to a method in expression tree also started executing those tasks. Fix is to rather than passing task array, create task lazily by passing func. So when we call await, we created task then only.

I ran SqlServer.FunctionalTests (which takes about 2-3 mins) in a loop for about an hour. No failures. I saw failures of this in local builds & CI too.
